### PR TITLE
Implement ICS-26 router for `RecvPacket` in `{Validation,Execution}Context`

### LINKED ISSUE #322: Make `ValidationContext` and `ExecutionContext` implement the ICS-26 router
When we wrote ADR 5, we forgot to consider how the ICS-26 router would work with the new `ValidationContext` and `ExecutionContext`. Also, since then, the removal of crossing hellos from connection and channel handshakes changed how certain handlers interact with the router (e.g. we can no longer emit events directly in `ChanOpenTry` and `ChanOpenAck` handlers). This issue serves to amend ADR 5 in order to address all of the above.

The bottom line is: both `validate()` and `execute()` need to perform the relevant parts of the ICS-26 router. As an example, here is what `validate()` and `execute()` would do to handle a `ChanOpenTry` message:

+ `validate()`:
    1. Get the associated `Module` based on the `msg.port_id`
    2. Call the core `chan_open_try::validate()` handler
    3. Query the host for the `channel_id` (through the channel counter)
    4. Call the `Module` callback `on_chan_open_try_validate()` (which only performs validation)

+ `execute()`:
    1. Get the associated `Module` based on the `msg.port_id`
    2. Query the host for the `channel_id` (through the channel counter)
    3. Call the `Module` callback `on_chan_open_try_execute(channel_id, ...)` (which only performs state changes), and save the returned `version`
    4. Call the core `chan_open_try::execute(version, ...)` handler, feeding it the `version` obtained from the `execute()` callback

A few things to note:
+ we need to extend the [`Module` trait](https://github.com/cosmos/ibc-rs/blob/968e553605acacdf799f51745373e08e02365a3b/crates/ibc/src/core/ics26_routing/context.rs#L113) with validation methods (1 per callback).
+ The `channel_id` is no longer generated by the core handler, as specified in the specs (although see [this spec issue](https://github.com/cosmos/ibc/issues/861) which shows that the spec needs to be updated). Instead we generate it in what is effectively the ICS-26 router.

### PR DESCRIPTION
**Pull request by plafer**
Work towards: #322

Changed the scope of this PR to include only `recvPacket`, since it turned out to have many subtleties to get right.

## Description

As with the other PRs, this one focuses on correctness. The more we get closer to being done with #322, the more it becomes obvious that we need a refactor. Let's hold off until we're completely done though; it will be easier to make a big refactor after #279.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs [hunk 1]
```diff
@@ -56,10 +56,6 @@ pub fn process<Ctx: ChannelReader>(
         Err(e) => return Err(e),
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs:56]: There is a slight deviation in the way we handle the `PacketAcknowledgementNotFound` error variant. We perceive it as an error, but eventually, it is accepted as a valid arm. I think `get_packet_acknowledgement(...)` should return `Result<Option<AcknowledgementCommitment>, PacketError>` type, so under normal circumstance, the `None` variant would show a valid absence of ACK, but in case where ever needed as an error, we can use another match or apply `ok_or()` (Though I didn't see any)

File: crates/ibc/src/mock/client_state.rs [hunk 2]
```diff
@@ -355,6 +355,22 @@ impl ClientState for MockClientState {
         Ok(())
     }
 
+    #[cfg(feature = "val_exec_ctx")]
+    fn new_verify_packet_data(
```
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/client_state.rs:359]: So, this should be renamed to `verify_packet_data` eventually. Okay

File: crates/ibc/src/core/ics02_client/client_state.rs [hunk 3]
```diff
@@ -175,6 +175,22 @@ pub trait ClientState:
         expected_client_state: Any,
     ) -> Result<(), ClientError>;
 
+    /// Verify a `proof` that a packet has been commited.
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/client_state.rs]: Typo from past: "committed"

File: crates/ibc/src/core/ics02_client/client_state.rs [hunk 4]
```diff
@@ -175,6 +175,22 @@ pub trait ClientState:
...
+        connection_end: &ConnectionEnd,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+        sequence: Sequence,
+        commitment: PacketCommitment,
+    ) -> Result<(), ClientError>;
+
     /// Verify a `proof` that a packet has been commited.
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/client_state.rs]: Typo from past: "committed"

File: crates/ibc/src/applications/transfer/relay/on_recv_packet.rs [hunk 5]
```diff
@@ -60,3 +60,78 @@ pub fn process_recv_packet<Ctx: 'static + TokenTransferContext>(
...
+#[cfg(feature = "val_exec_ctx")]
+mod val_exec_ctx {
+    pub use super::*;
+    use crate::core::ics04_channel::handler::ModuleExtras;
+
+    pub fn process_recv_packet_execute<Ctx: TokenTransferContext>(
+        ctx: &mut Ctx,
+        packet: &Packet,
+        data: PacketData,
+    ) -> Result<ModuleExtras, (ModuleExtras, TokenTransferError)> {
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/relay/on_recv_packet.rs:75]: Are we trying to move away from the output argument?

One thing that occurred to me is that grouping two orthogonal events and logs under the same struct (ModuleExtras) doesn't seem right.

- A logger should work till the last line of code is worked, but it's not true for events. Hence, the logger handler should be passed to functions somehow similar to the output arg.
- Contrary to logs, events can be returned or produced by functions, and this should be reflected in results.

I think this should be related to #184

File: crates/ibc/src/applications/transfer/acknowledgement.rs [hunk 6]
```diff
@@ -56,17 +56,21 @@ impl Display for TokenTransferAcknowledgement {
 
 impl From<TokenTransferAcknowledgement> for Vec<u8> {
     fn from(ack: TokenTransferAcknowledgement) -> Self {
+        // WARNING: Make sure all branches always return a non-empty vector.
+        // Otherwise, the conversion to `Acknowledgement` will panic.
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/acknowledgement.rs:60]: Would this panic, as it wasn't before?

Imho, this warning sounds like something couldn't be handled with the Rust error system, so it's up to users to be cautious about it. This also led to using `expect` in `From<TokenTransferAcknowledgement>` conversion.

We moved the `is_empty` check from the `write_acknowledgement::process` level to the higher `on_recv_packet` level. This makes me think we should change the return of `on_recv_packet` to `Result<Acknowledgment, PacketError>`. not sure ofc.

I did like removing that `is_empty()` check, but I have no idea how that could be handled better :)

File: crates/ibc/src/core/context.rs [hunk 7]
```diff
@@ -1114,4 +1169,121 @@ mod val_exec_ctx {
...
+                    // the sequence number has already been incremented, so
+                    // another relayer already relayed the packet
+                    msg.packet.sequence < next_seq_recv
+                }
+            };
+
+            if packet_already_received {
+                return Ok(());
+            }
+        }
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/context.rs:1219]: This sounds more like a validation step.

**Review by Farhad-Shabani — COMMENTED** [review]
All good :ok_hand: 
I just left some thoughts to get also your take on them

File: crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs (hunk 1 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs:56]: Note that this is legacy code and is never called from `Validation/ExecutionContext`. We no longer treat `PacketAcknowledgementNotFound` specially in the [new implementation](https://github.com/cosmos/ibc-rs/blob/dc84bf2e8359f4bfcb78e3a8cbb9cad3a1c47002/crates/ibc/src/core/ics04_channel/handler/recv_packet.rs#L153-L161).

> I think get_packet_acknowledgement(...) should return Result<Option<AcknowledgementCommitment>, PacketError> type, so under normal circumstance, the None variant would show a valid absence of ACK

What's a "valid" absence of an ACK (as opposed to an "invalid one")? The (undocumented) contract of `get_packet_acknowledgement()` is "return an ack commitment, or tell me why you were not able to do it". The acknowledgement not being present is just another reason why the host would not be able to return it, and is an error in this context. IMO `Result<Option<AcknowledgementCommitment>, PacketError>` is harder to read, and doesn't provide us any gains that I can see. Note also that this would make the new implementation more complex, as we'd not only need to check whether or not there was an error, but also if the return was `None` in the success case.

File: crates/ibc/src/core/ics02_client/client_state.rs (hunk 3 above)
Comment (plafer) [reply crates/ibc/src/core/ics02_client/client_state.rs]: [147ec47](https://github.com/cosmos/ibc-rs/pull/377/commits/147ec47a78503d30dfb26cdc1cba20766066b9c0)

File: crates/ibc/src/core/ics02_client/client_state.rs (hunk 4 above)
Comment (plafer) [reply crates/ibc/src/core/ics02_client/client_state.rs]: [147ec47](https://github.com/cosmos/ibc-rs/pull/377/commits/147ec47a78503d30dfb26cdc1cba20766066b9c0)

File: crates/ibc/src/applications/transfer/relay/on_recv_packet.rs (hunk 5 above)
Comment (plafer) [reply crates/ibc/src/applications/transfer/relay/on_recv_packet.rs:75]: > Are we trying to move away from the output argument?

Yes.

> A logger should work till the last line of code is worked, but it's not true for events. Hence, the logger handler should be passed to functions somehow similar to the output arg.

Actually the events also need to be returned even in the case of an error (see [ibc-go](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/apps/transfer/ibc_module.go#L176-L210)).

This interface looks funny because I wanted to still use `?` for errors, but then `ModuleExtras` need to be returned also in the case of an error. `(ModuleExtras, Result<(), Error>)` would not have allowed the use of `?`.

There's probably a cleaner way to do it, but again, I prioritized correctness.

File: crates/ibc/src/applications/transfer/acknowledgement.rs (hunk 6 above)
Comment (plafer) [reply crates/ibc/src/applications/transfer/acknowledgement.rs:60]: I'm also not a fan of needing that comment. Note also that I added a test to check that this conversion doesn't return a non-empty vector.

The fix would look like:
```rust
impl From<TokenTransferAcknowledgement> for NonEmptyVec<u8> {
    // ...
}

impl From<TokenTransferAcknowledgement> for Acknowledgement {
    fn from(ack: TokenTransferAcknowledgement) -> Self {
        let v: NonEmptyVec<u8> = ack.into();

        // this can be an `into()` because the only possible error case
        // (empty vector) is not possible with `NonEmptyVec<u8>`
        v.into()
    }
}
```

Comment (plafer) [reply crates/ibc/src/applications/transfer/acknowledgement.rs:60]: Created #389

File: crates/ibc/src/core/context.rs (hunk 7 above)
Comment (plafer) [reply crates/ibc/src/core/context.rs:1219]: It's not; notice that what we're trying to achieve here seeing if we should return early or not (or, in other words, don't do any writes). Does that make sense?

File: crates/ibc/src/applications/transfer/acknowledgement.rs (hunk 6 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/applications/transfer/acknowledgement.rs:60]: :pray:

File: crates/ibc/src/applications/transfer/relay/on_recv_packet.rs (hunk 5 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/applications/transfer/relay/on_recv_packet.rs:75]: > Actually the events also need to be returned even in the case of an error (see ibc-go). This interface looks funny because I wanted to still use ? for errors, but then ModuleExtras need to be returned also in the case of an error. (ModuleExtras, Result<(), Error>) would not have allowed the use of ?.

Yea, I also tried that option locally, but it does not buy much. I see.

File: crates/ibc/src/core/context.rs (hunk 7 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:1219]: :thinking: I still think it is not of an early return type. A duplicate msg is not valid to process. A case can be considered: multiple relayers try to manipulate and attack by sending tons of messages, none of them should be allowed to pass the validation.

File: crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/core/ics04_channel/handler/write_acknowledgement.rs:56]: > Note that this is legacy code and is never called from Validation/ExecutionContext. We no longer treat PacketAcknowledgementNotFound specially in the new implementation.

So, if it would not be the case, we shoule be ok then. thank you :pray:

File: crates/ibc/src/core/context.rs (hunk 7 above)
Comment (plafer) [reply crates/ibc/src/core/context.rs:1219]: > A duplicate msg is not valid to process.

Actually it is. With every packet to relay, relayers are in a race to relay the packet, and they'll all try to send the same `recvPacket` transaction at the same time. But only the first one will be accepted by the chain. Previously, ibc-go didn't have that "already received" check, and so all the relayers' transactions (except the first successful one) would error, which meant that other messages in the transaction would be discarded, and increased fees for the relayer (I think). But there's no point in penalizing relayers in this way, so ibc-go made this case a valid `recvPacket` message. See this ibc-go issue for more context: https://github.com/cosmos/ibc-go/issues/245.

> multiple relayers try to manipulate and attack by sending tons of messages

How is this an attack? What exactly would be manipulated? AFAIK, they're not gaining anything, because they spent money on the transaction, but the transaction ended up being a no-op and not changing any of the blockchain state.

Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:1219]: >  Previously, ibc-go didn't have that "already received" check, and so all the relayers' transactions (except the first successful one) would error, 

I don't see it was considered as a valid msg. Instead of returning errors they return no-ops. 
Let's look at it from another side: which of the execution steps (state change, store, emit, log) are done on a redundant message? All of them are dropped as they are "not valid for execution"

(I am discussing this to clarify the validation and execution boundaries. All in all, It's not a big deal)

> How is this an attack? What exactly would be manipulated? AFAIK, they're not gaining anything, because they spent money on the transaction, but the transaction ended up being a no-op and not changing any of the blockchain state.

Something like DDoS attack, Getting the network down for a while by consuming the chain or relayers processing capacity to incur a hack by manipulating some other parts at that time.

Comment (plafer) [reply crates/ibc/src/core/context.rs:1219]: > Let's look at it from another side: which of the execution steps (state change, store, emit, log) are done on a redundant message? All of them are dropped as they are "not valid for execution"

That's right; none are executed. Stated differently, sending a `recvPacket` msg for a packet that has already been received is a valid transaction which doesn't change the state. The reason for it to be a valid message is simply so that relayers don't have to pay extra fees for doing exactly what they're supposed to do. If we failed the validation, then they would pay extra fees; so instead we succeed the transaction, but don't change the state machine.

> Something like DDoS attack, Getting the network down for a while by consuming the chain or relayers processing capacity to incur a hack by manipulating some other parts at that time.

Can you expand on how that attack would work? How is this any different than someone spamming the chain with any other valid transaction (e.g. sending tokens back and forth between their own accounts)?

Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:1219]: Per our discussion and giving it more thoughts, it makes sense :pray:

**Review by Farhad-Shabani — APPROVED** [review]
